src/main/java/net/fabricmc/loader/api/VersionRange.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.api;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+
+public class VersionRange {
+	public enum Type {
+		INVALID(version -> "unknown version"),
```

Comment:
I'm pretty sure there's currently no way to get an actual `INVALID` range without messing with the program itself, since I treat anything that doesn't have a semantic version prefix as string version equality.
I *do* think it's a good idea to keep an "oh god oh fuck" value that *isn't* a potentially-crashing null, though.

Reply:
So `INVALID` was only used for `"*moretext"` type situations. Those now just generate a standard `ANY` comparison, if they're supposed to be equality shout at me.